### ISSUE TITLE
Xiaomi MiJia wireless switch - "false click"

### ISSUE THREAD

**Issue by MiloshCZ** [stas-demydiuk/domoticz-zigbee2mqtt-plugin#14]
Hi,
I have found, that there is "false click" because Xiaomi switch sends heartbeat every hour.

HeartBeat message:
```
{
  "battery" : "100.00",
  "voltage" : 3042,
  "linkquality" : 115,
  "device" : {
    "ieeeAddr" : "0x00158d000228709c",
    "friendlyName" : "0x00158d000228709c",
    "type" : "EndDevice",
    "nwkAddr" : 29140,
    "manufId" : 4151,
    "manufName" : "LUMI",
    "powerSource" : "Battery",
    "modelId" : "lumi.sensor_switch",
    "status" : "online"
  }
}
```
Domoticz log:
```
2018-10-17 16:09:19.612 (Zigbee2MQTT) 30 32 32 38 37 30 39 63 22 2c 22 66 72 69 65 6e 64 6c 79 4e 0228709c","friendlyN
2018-10-17 16:09:19.612 (Zigbee2MQTT) 61 6d 65 22 3a 22 30 78 30 30 31 35 38 64 30 30 30 32 32 38 ame":"0x00158d000228
2018-10-17 16:09:19.612 (Zigbee2MQTT) 37 30 39 63 22 2c 22 74 79 70 65 22 3a 22 45 6e 64 44 65 76 709c","type":"EndDev
2018-10-17 16:09:19.612 (Zigbee2MQTT) 69 63 65 22 2c 22 6e 77 6b 41 64 64 72 22 3a 32 39 31 34 30 ice","nwkAddr":29140
2018-10-17 16:09:19.612 (Zigbee2MQTT) 2c 22 6d 61 6e 75 66 49 64 22 3a 34 31 35 31 2c 22 6d 61 6e ,"manufId":4151,"man
2018-10-17 16:09:19.612 (Zigbee2MQTT) 75 66 4e 61 6d 65 22 3a 22 4c 55 4d 49 22 2c 22 70 6f 77 65 ufName":"LUMI","powe
2018-10-17 16:09:19.613 (Zigbee2MQTT) 72 53 6f 75 72 63 65 22 3a 22 42 61 74 74 65 72 79 22 2c 22 rSource":"Battery","
2018-10-17 16:09:19.613 (Zigbee2MQTT) 6d 6f 64 65 6c 49 64 22 3a 22 6c 75 6d 69 2e 73 65 6e 73 6f modelId":"lumi.senso
2018-10-17 16:09:19.613 (Zigbee2MQTT) 72 5f 73 77 69 74 63 68 22 2c 22 73 74 61 74 75 73 22 3a 22 r_switch","status":"
2018-10-17 16:09:19.613 (Zigbee2MQTT) 6f 6e 6c 69 6e 65 22 7d 7d .. .. .. .. .. .. .. .. .. .. .. online"}}
2018-10-17 16:09:19.614 (Zigbee2MQTT) MQTT message: zigbee2mqtt/0x00158d000228709c {'voltage': 3042, 'device': {'type': 'EndDevice', 'powerSource': 'Battery', 'friendlyName': '0x00158d000228709c', 'status': 'online', 'manufName': 'LUMI', 'nwkAddr': 29140, 'modelId': 'lumi.sensor_switch', 'ieeeAddr': '0x00158d000228709c', 'manufId': 4151}, 'battery': '100.00', 'linkquality': 115}
2018-10-17 16:09:19.616 (Spínač) Updating device from 1:'10' to have values 1:'10'.
```

Click message:
```
{
  "battery" : "100.00",
  "voltage" : 3042,
  "linkquality" : 126,
  "click" : "single",
  "device" : {
    "ieeeAddr" : "0x00158d000228709c",
    "friendlyName" : "0x00158d000228709c",
    "type" : "EndDevice",
    "nwkAddr" : 29140,
    "manufId" : 4151,
    "manufName" : "LUMI",
    "powerSource" : "Battery",
    "modelId" : "lumi.sensor_switch",
    "status" : "online"
  }
}
```
Domoticz Log
```
2018-10-17 16:10:01.764 (Zigbee2MQTT) 30 c3 02 00 1e 7a 69 67 62 65 65 32 6d 71 74 74 2f 30 78 30 0�...zigbee2mqtt/0x0
2018-10-17 16:10:01.764 (Zigbee2MQTT) 30 31 35 38 64 30 30 30 32 32 38 37 30 39 63 7b 22 62 61 74 0158d000228709c{"bat
2018-10-17 16:10:01.764 (Zigbee2MQTT) 74 65 72 79 22 3a 22 31 30 30 2e 30 30 22 2c 22 76 6f 6c 74 tery":"100.00","volt
2018-10-17 16:10:01.764 (Zigbee2MQTT) 61 67 65 22 3a 33 30 34 32 2c 22 6c 69 6e 6b 71 75 61 6c 69 age":3042,"linkquali
2018-10-17 16:10:01.765 (Zigbee2MQTT) 74 79 22 3a 31 32 36 2c 22 63 6c 69 63 6b 22 3a 22 73 69 6e ty":126,"click":"sin
2018-10-17 16:10:01.765 (Zigbee2MQTT) 67 6c 65 22 2c 22 64 65 76 69 63 65 22 3a 7b 22 69 65 65 65 gle","device":{"ieee
2018-10-17 16:10:01.765 (Zigbee2MQTT) 41 64 64 72 22 3a 22 30 78 30 30 31 35 38 64 30 30 30 32 32 Addr":"0x00158d00022
2018-10-17 16:10:01.765 (Zigbee2MQTT) 38 37 30 39 63 22 2c 22 66 72 69 65 6e 64 6c 79 4e 61 6d 65 8709c","friendlyName
2018-10-17 16:10:01.765 (Zigbee2MQTT) 22 3a 22 30 78 30 30 31 35 38 64 30 30 30 32 32 38 37 30 39 ":"0x00158d000228709
2018-10-17 16:10:01.765 (Zigbee2MQTT) 63 22 2c 22 74 79 70 65 22 3a 22 45 6e 64 44 65 76 69 63 65 c","type":"EndDevice
2018-10-17 16:10:01.765 (Zigbee2MQTT) 22 2c 22 6e 77 6b 41 64 64 72 22 3a 32 39 31 34 30 2c 22 6d ","nwkAddr":29140,"m
2018-10-17 16:10:01.765 (Zigbee2MQTT) 61 6e 75 66 49 64 22 3a 34 31 35 31 2c 22 6d 61 6e 75 66 4e anufId":4151,"manufN
2018-10-17 16:10:01.765 (Zigbee2MQTT) 61 6d 65 22 3a 22 4c 55 4d 49 22 2c 22 70 6f 77 65 72 53 6f ame":"LUMI","powerSo
2018-10-17 16:10:01.766 (Zigbee2MQTT) 75 72 63 65 22 3a 22 42 61 74 74 65 72 79 22 2c 22 6d 6f 64 urce":"Battery","mod
2018-10-17 16:10:01.766 (Zigbee2MQTT) 65 6c 49 64 22 3a 22 6c 75 6d 69 2e 73 65 6e 73 6f 72 5f 73 elId":"lumi.sensor_s
2018-10-17 16:10:01.766 (Zigbee2MQTT) 77 69 74 63 68 22 2c 22 73 74 61 74 75 73 22 3a 22 6f 6e 6c witch","status":"onl
2018-10-17 16:10:01.766 (Zigbee2MQTT) 69 6e 65 22 7d 7d .. .. .. .. .. .. .. .. .. .. .. .. .. .. ine"}}
2018-10-17 16:10:01.767 (Zigbee2MQTT) MQTT message: zigbee2mqtt/0x00158d000228709c {'voltage': 3042, 'click': 'single', 'device': {'type': 'EndDevice', 'powerSource': 'Battery', 'friendlyName': '0x00158d000228709c', 'status': 'online', 'manufName': 'LUMI', 'nwkAddr': 29140, 'modelId': 'lumi.sensor_switch', 'ieeeAddr': '0x00158d000228709c', 'manufId': 4151}, 'battery': '100.00', 'linkquality': 126}
2018-10-17 16:10:01.768 (Spínač) Updating device from 1:'10' to have values 1:'10'.
```

**Comment by Tezzlicious** [comment]
It's for sure false clicking. Here is a snippet of my domoticz log:




2018-10-20 20:37:04 | Double Click
2018-10-20 19:36:46 | Double Click
2018-10-20 18:36:27 | Double Click
2018-10-20 17:36:05 | Double Click
2018-10-20 16:35:39 | Double Click
2018-10-20 15:35:19 | Double Click
2018-10-20 14:35:04 | Double Click
2018-10-20 13:34:33 | Double Click
2018-10-20 12:34:13 | Double Click
2018-10-20 11:33:51 | Double Click
2018-10-20 10:33:33 | Double Click
2018-10-20 09:33:15 | Double Click
2018-10-20 08:32:55 | Double Click
2018-10-20 07:32:33 | Double Click
2018-10-20 07:32:32 | Double Click
2018-10-20 06:35:16 | Triple Click
2018-10-20 05:34:58 | Triple Click
2018-10-20 04:34:38 | Triple Click
2018-10-20 03:34:19 | Triple Click
2018-10-20 02:34:04 | Triple Click
2018-10-20 01:33:37 | Triple Click
2018-10-20 01:29:54 | Triple Click
2018-10-20 00:41:01 | Click
2018-10-19 23:40:41 | Click
2018-10-19 22:40:21 | Click
2018-10-19 21:39:59 | Click
2018-10-19 20:39:40 | Click
2018-10-19 19:39:20 | Click
2018-10-19 18:38:58 | Click
2018-10-19 17:38:34 | Click
2018-10-19 16:38:12 | Click
2018-10-19 15:37:51 | Click
2018-10-19 14:37:30

**Comment by stas-demydiuk** [comment]
Guys, @MiloshCZ @Tezzlicious what do you think is the best solution for that, I see such options:
1. ignore all device messages if current device state in domoticz equals state in the MQTT message but in this case we will skip all heartbeat messages and messages with the same device state (like couple Clicks in a row) until device be switched in a new state
2. ignore just all heartbeat messages (messages without device state)
3. manually set the state to "Off" after some time, as an option using events editor as all these switches more similar to push-on button than to regular switch

**Comment by Tezzlicious** [comment]
I would pick option 2.

**Comment by MiloshCZ** [comment]
1. NO - this will ignore clicks
2. YES - But there is small issue, that there is no update about battery and signal level in case of long inactivity
3. YES - this solution has benefit, that signal and batteryy level is updated periodicaly.

Or you can send something like this:
`.../json.htm?type=command&param=udevice&nvalue=0&idx=41&rssi=10&battery=89`
It will update everything, but in log there is no action
![image](https://user-images.githubusercontent.com/13035760/47450726-4707f380-d7c6-11e8-81ca-665ca5c8fb87.png)

**Comment by Tezzlicious** [comment]
So that command in combination with option 2. Heartbeats update values only, no trigger.

**Comment by stas-demydiuk** [comment]
@MiloshCZ @Tezzlicious the issue should be fixed with plugin in the latest master branch

**Comment by MarcE85** [comment]
If i update to this version i cant use my xiaomi switches. Domoticz / DzVents only reacts to the first "Click" Command. After that i have to do a "Dobule Click" and then a "Click" again to generate an Event in DzVents.

Regards Marc

**Comment by stas-demydiuk** [comment]
@MarcE85 Yes, I have the same issue with my scripts, and I manually set device status to 'off' as workaround. Below is the example of dzVents script:

```
return {
	active = true,

	on = {
		devices = {
			'mySwitch',
		}
	},

	execute = function(domoticz, device)
		if (device.state == 'Switch 1') then
			device.switchOff()
			domoticz.devices('Light1').toggleSwitch()
		end
		
		if (device.state == 'Switch 2') then
			device.switchOff()
			domoticz.devices('Light2').toggleSwitch()
		end
		
		if (device.state == 'Both_Click') then
			device.switchOff()
			domoticz.devices('Light3').toggleSwitch()
		end
	end
}
```

**Comment by wifiminime** [comment]
Hi, 

This workaound does not work for me, i had already filed an issue for this...
https://github.com/stas-demydiuk/domoticz-zigbee2mqtt-plugin/issues/125

Any idea why your workaround does not work for me? 
I have an Philips Hue dimmer switch and i can only use the buttons once, very annoying.... have not found a way around this.
Looking forward to an update of the plugin, as you already mentioned, most wiresless switches are push button and not a regular switch (they have a 'state')

**Comment by stas-demydiuk** [comment]
The issue was fixed in scope of #125
